File: examples/agentic_patch_runner.py
from dataclasses import dataclass, replace
from pathlib import PurePosixPath
from time import monotonic
from typing import Protocol


WRITE_TOOLS = {"apply_patch", "write_file", "delete_file"}


@dataclass(frozen=True)
class ToolCall:
    name: str
    path: str
    arguments: dict[str, object]


@dataclass(frozen=True)
class Observation:
    succeeded: bool
    output: str


class Workspace(Protocol):
    @property
    def revision(self) -> str: ...

    def execute(self, call: ToolCall) -> Observation: ...


class Model(Protocol):
    def next_call(self, goal: str, transcript: list[tuple[ToolCall, Observation]]) -> ToolCall: ...


class AgenticPatchRunner:
    def __init__(self, workspace: Workspace, model: Model, turn_timeout_seconds: float = 30.0) -> None:
        self.workspace = workspace
        self.model = model
        self.turn_timeout_seconds = turn_timeout_seconds

    # @agent-context design
    # The runner asks for one tool call per model turn rather than accepting a batch. Tool output is part of the next decision, and batching encouraged the model to keep executing an obsolete plan after the first edit changed the workspace. Single-call turns also provide a precise cancellation boundary.
    def run(self, goal: str, allow_writes: bool) -> str:
        transcript: list[tuple[ToolCall, Observation]] = []
        expected_revision = self.workspace.revision

        while True:
            started_at = monotonic()
            call = self._normalize(self.model.next_call(goal, transcript))

            # @agent-context invariant
            # Check authorization after normalization and execute the exact checked value. Moving normalization below this branch could let aliases or relative path segments transform an apparently harmless request into a write outside the intended workspace boundary.
            if call.name in WRITE_TOOLS and not allow_writes:
                return "blocked: workspace writes were not authorized"

            # @agent-context dependency
            # Revision values come from the editor bridge and represent human activity, not merely file bytes. An edit followed by an undo still changes the revision because the human may have rejected the agent's direction. Do not replace this comparison with a content hash.
            if self.workspace.revision != expected_revision:
                return "blocked: workspace changed during the run"

            observation = self.workspace.execute(call)
            transcript.append((call, observation))
            expected_revision = self.workspace.revision

            # @agent-context history
            # Failed observations stay in the transcript. Before incident AGENT-42, failures were discarded and the model retried the same invalid rename until the turn budget was exhausted. Keeping the error lets the following turn choose a different action and makes replay diagnostics complete.
            if observation.succeeded and call.name == "finish":
                return observation.output

            if monotonic() - started_at > self.turn_timeout_seconds:
                return "blocked: model turn exceeded its time budget"

    # @agent-context compatibility
    # Prompt paths are persisted with POSIX separators so evaluation transcripts remain portable across operating systems. The concrete workspace adapter performs native-path conversion immediately before I/O. Changing the serialized form here would invalidate existing replay fixtures.
    def _normalize(self, call: ToolCall) -> ToolCall:
        normalized_path = str(PurePosixPath(call.path.replace("\\", "/")))
        return replace(call, path=normalized_path)

File: examples/agentic_patch_runner.rs
use std::time::{Duration, Instant};

#[derive(Clone)]
struct ToolCall {
    name: String,
    path: String,
}

struct Observation {
    succeeded: bool,
    output: String,
}

trait Workspace {
    fn revision(&self) -> String;
    fn execute(&mut self, call: &ToolCall) -> Observation;
}

trait Model {
    fn next_call(&mut self, goal: &str, transcript: &[(ToolCall, Observation)]) -> ToolCall;
}

struct AgenticPatchRunner<W, M> {
    workspace: W,
    model: M,
    turn_timeout: Duration,
}

impl<W: Workspace, M: Model> AgenticPatchRunner<W, M> {
    /* @agent-context design
     * Generate one tool call at a time. Batched calls proved unsafe once an early edit invalidated the model's remaining assumptions, while single-call turns provide deterministic checkpoints for cancellation, approval, and replay.
     */
    fn run(&mut self, goal: &str, allow_writes: bool) -> String {
        let mut transcript = Vec::new();
        let mut expected_revision = self.workspace.revision();

        loop {
            let started_at = Instant::now();
            let proposed = self.model.next_call(goal, &transcript);
            let call = Self::normalize(proposed);

            /* @agent-context invariant
             * Check permission on the normalized value and pass that exact value to the executor. Normalizing after authorization could let a relative path or alias change the operation that was actually approved.
             */
            if call.name == "apply_patch" && !allow_writes {
                return "blocked: workspace writes were not authorized".into();
            }

            /* @agent-context dependency
             * Revisions are activity tokens supplied by the editor bridge rather than content hashes. An edit followed by undo still advances the token because it records that a human reviewed and rejected a possible direction.
             */
            if self.workspace.revision() != expected_revision {
                return "blocked: workspace changed during the run".into();
            }

            let observation = self.workspace.execute(&call);
            let finished = observation.succeeded && call.name == "finish";
            let output = observation.output.clone();
            transcript.push((call, observation));
            expected_revision = self.workspace.revision();

            /* @agent-context history
             * Failed observations remain in the transcript. Incident AGENT-42 discarded executor errors, causing the model to retry the same invalid rename until its budget expired and leaving replay diagnostics without the triggering failure.
             */
            if finished {
                return output;
            }

            if started_at.elapsed() > self.turn_timeout {
                return "blocked: model turn exceeded its time budget".into();
            }
        }
    }

    /* @agent-context compatibility
     * Persist slash-separated prompt paths on every platform because evaluation transcripts are shared across operating systems. The workspace implementation converts them to native paths only at the filesystem boundary.
     */
    fn normalize(mut call: ToolCall) -> ToolCall {
        call.path = call.path.replace('\\', "/");
        call
    }
}

fn main() {}
